### PR TITLE
Insert JsDoc comment templates for additional nodes

### PR DESCRIPTION
**Pull request by uniqueiniquity**
In response to https://developercommunity.visualstudio.com/content/problem/138023/typescript-editor-jsdoc-comments-not-auto-generate.html.

Previously, we were not inserting jsDoc comment templates before interface declarations, method signatures, property signatures, and enums, whereas we were for class declarations and method declarations. This adds this functionality in.

Also, following a suggestion by @andy-ms, we no longer return undefined if it is inappropriate to insert a doc comment template. Instead, we return an empty template.

### CODE REVIEW

File: tests/cases/fourslash/docCommentTemplateInterfaces.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+////    /*methodbaz*/
+////    baz(message: any): void;
+////}
+
+verify.docCommentTemplateAt("interfaceFoo", /*expectedOffset*/ 8,
+`/**
+ * 
+ */`);
+
+verify.emptyDocCommentTemplateAt("propertybar");
```
Comment (andy-ms) [inline tests/cases/fourslash/docCommentTemplateInterfaces.ts]: Why not give us the above template in this case too?

Comment (uniqueiniquity) [reply tests/cases/fourslash/docCommentTemplateInterfaces.ts]: I can, it was just outside of the scope of this particular bug fix. I can go and add these too if we want.

Comment (andy-ms) [reply tests/cases/fourslash/docCommentTemplateInterfaces.ts]: I think we could just return that template (with the current indentation) at any arbitrary location where we don't have specific handling.

File: src/services/jsDoc.ts [hunk 2]
```diff
@@ -256,18 +260,19 @@ namespace ts.JsDoc {
...
                 case SyntaxKind.FunctionDeclaration:
                 case SyntaxKind.MethodDeclaration:
                 case SyntaxKind.Constructor:
-                    const { parameters } = commentOwner as FunctionDeclaration | MethodDeclaration | ConstructorDeclaration;
+                case SyntaxKind.MethodSignature:
+                    const { parameters } = commentOwner as FunctionDeclaration | MethodDeclaration | ConstructorDeclaration | MethodSignature;
                     return { commentOwner, parameters };
 
                 case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.InterfaceDeclaration:
```
Comment (andy-ms) [inline src/services/jsDoc.ts]: Looks like `commentOwner` is only used in one place: `if (commentOwner.getStart() < position) { return undefined; }`.
Seems like this function could just be `getParameters`...

File: tests/cases/fourslash/docCommentTemplateInterfaces.ts (hunk 1 above)
Comment (uniqueiniquity) [reply tests/cases/fourslash/docCommentTemplateInterfaces.ts]: Any arbitrary location? I feel like it should always be leading up to a declaration or member of some sort.

File: src/services/jsDoc.ts (hunk 2 above)
Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: I like this current setup since we have an explicit list of what places are permissible for these templates, and it's easy to add new things.

File: src/services/jsDoc.ts [hunk 3]
```diff
@@ -188,24 +190,26 @@ namespace ts.JsDoc {
      * be performed.
      */
     export function getDocCommentTemplateAtPosition(newLine: string, sourceFile: SourceFile, position: number): TextInsertion {
+        const emptyDocComment = { newText: "", caretOffset: 0 };
+
         // Check if in a context where we don't want to perform any insertion
         if (isInString(sourceFile, position) || isInComment(sourceFile, position) || hasDocComment(sourceFile, position)) {
-            return undefined;
+            return emptyDocComment;
```
Comment (mhegazy) [inline src/services/jsDoc.ts]: so why do we return an empty object here? what is wrong in undefined?

File: src/services/jsDoc.ts (hunk 2 above)
Comment (andy-ms) [reply src/services/jsDoc.ts]: If a user types `/**` in an empty file and hits enter, I would expect to see an empty doc comment -- it's unambiguous that that's what they're trying to do. No sense in not giving the user formatting just because they're not above a declaration.

File: src/services/jsDoc.ts [hunk 4]
```diff
@@ -255,19 +259,21 @@ namespace ts.JsDoc {
...
-                    const { parameters } = commentOwner as FunctionDeclaration | MethodDeclaration | ConstructorDeclaration;
+                case SyntaxKind.MethodSignature:
+                    const { parameters } = commentOwner as FunctionDeclaration | MethodDeclaration | ConstructorDeclaration | MethodSignature;
                     return { commentOwner, parameters };
 
                 case SyntaxKind.ClassDeclaration:
+                case SyntaxKind.InterfaceDeclaration:
+                case SyntaxKind.PropertySignature:
+                case SyntaxKind.EnumDeclaration:
+                case SyntaxKind.EnumMember:
```
Comment (mhegazy) [inline src/services/jsDoc.ts]: type alias as well.

File: src/services/jsDoc.ts (hunk 3 above)
Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: Was @andy-ms's suggestion. Andy, could you elaborate?

File: src/services/jsDoc.ts [hunk 5]
```diff
@@ -255,19 +259,21 @@ namespace ts.JsDoc {
     }
     function getCommentOwnerInfo(tokenAtPos: Node): CommentOwnerInfo | undefined {
         // TODO: add support for:
-        // - enums/enum members
-        // - interfaces
-        // - property declarations
         // - potentially property assignments
```
Comment (mhegazy) [inline src/services/jsDoc.ts]: do not think this is comment is needed either. we handle these in `BinaryExpression` section.

File: src/services/jsDoc.ts (hunk 3 above)
Comment (andy-ms) [reply src/services/jsDoc.ts]: See https://github.com/Microsoft/TypeScript/pull/19544#discussion_r147851114

File: src/services/jsDoc.ts (hunk 2 above)
Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: Even if they're in the middle of an expression? For example, what if someone's adding a comment to a parameter (`foo(/*param*/ undefined)`) and then accidentally hit * twice. I guess they can just hit undo but it seems easy to trigger.

File: src/services/jsDoc.ts (hunk 3 above)
Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: aha, I misunderstood what you meant by empty doc comment template.

File: src/services/jsDoc.ts (hunk 2 above)
Comment (andy-ms) [reply src/services/jsDoc.ts]: Do we fill in the comment immediately on the second `*` or do we wait for them to hit enter?

Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: It's immediate.

Comment (andy-ms) [reply src/services/jsDoc.ts]: In vscode you must hit enter. What if a user wants to add a single-line comment to an interface?

Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: I mean same as my description above: if they accidentally hit a second star, then it will trigger, but otherwise it won't. Seems somehow less bad then triggering in the middle of a parenthetical block.

Comment (andy-ms) [reply src/services/jsDoc.ts]: I mean a single-line jsdoc comment. These seem pretty common.
```ts
/** My interface */
interface I {
    /** My method */
    m(): void;
}
```

Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: Huh. Yeah we never use that with the current implementation.

Comment (uniqueiniquity) [reply src/services/jsDoc.ts]: So maybe we should just autocomplete to empty, one-line jsDoc comments and then if there are params, autocomplete to a multi-line and fill in the params?

File: src/services/jsDoc.ts (hunk 3 above)
Comment (mhegazy) [reply src/services/jsDoc.ts]: so how is the empty file case related to `isInString` or `isIncomment` ?

Comment (mhegazy) [reply src/services/jsDoc.ts]: if you want to enable them all the time that is fine, but not in strings and not in comments

**Comment by andy-ms** [comment]
I feel like commenting

**Comment by uniqueiniquity** [comment]
I got rid of the empty template approach. Turns out it makes VS behave weirdly, whereas `undefined` induces the intended behavior.

**Comment by uniqueiniquity** [comment]
This function will now:
- return a single-line JsDoc comment unless there are parameters to document, in which case it will fill them in as before.
- autocomplete everywhere except for inside strings, comments, template literals, regex literals, and JSXText

File: src/services/jsDoc.ts [hunk 6]
```diff
@@ -201,13 +199,25 @@ namespace ts.JsDoc {
...
+        if (commentOwner.kind === SyntaxKind.JsxText) {
             return undefined;
         }
 
+        if (commentOwner.getStart() < position) {
+            // if climbing the tree found a declaration with parameters but the request was made inside it, complete to a single line comment
+            return singleLineTemplate;
+        }
+
+        if (parameters.length === 0) {
```
Comment (andy-ms) [inline src/services/jsDoc.ts]: Nit: Use `||` to combine with above

File: src/services/jsDoc.ts [hunk 7]
```diff
@@ -216,18 +226,16 @@ namespace ts.JsDoc {
...
-                    "param" + i;
-                if (isJavaScriptFile) {
-                    docParams += `${indentationStr} * @param {any} ${paramName}${newLine}`;
-                }
-                else {
-                    docParams += `${indentationStr} * @param ${paramName}${newLine}`;
-                }
+        for (let i = 0; i < parameters.length; i++) {
+            const currentName = parameters[i].name;
+            const paramName = currentName.kind === SyntaxKind.Identifier ?
```
Comment (andy-ms) [inline src/services/jsDoc.ts]:
```ts
const docParams = parameters.map(({name}, i) => {
    const nameText = isIdentifier(name) ? name.text : `param${i}`;
    const type = isJavaScriptFile ? "{any} " : "";
    return `${indentationStr} * @param ${type}${nameText}${newLine}`;
}).join();
```

File: tests/cases/fourslash/docCommentTemplateClassDeclMethods01.ts [hunk 8]
```diff
@@ -1,7 +1,7 @@
 /// <reference path='fourslash.ts' />
 
 const enum Indentation {
-    Standard = 8,
+    Standard = 3,
```
Comment (andy-ms) [inline tests/cases/fourslash/docCommentTemplateClassDeclMethods01.ts]: Why this change? It's unusual to set indentation to 3.

Comment (uniqueiniquity) [reply tests/cases/fourslash/docCommentTemplateClassDeclMethods01.ts]: Indentation is a really terrible name for it. It's actually the caret offset into the template. But indentation is the term consistently used for it. I could go in and change it all if warranted.

Comment (andy-ms) [reply tests/cases/fourslash/docCommentTemplateClassDeclMethods01.ts]: Maybe in the old version this actually was an indentation?
This really makes no sense as an enum and could just be `const`s.
